python/pylibcugraph/pylibcugraph/sssp.pyx

```diff
@@ -0,0 +1,249 @@
+    distance for each vertex from the source vertex, and the third item is a
+    device array containing the vertex identifier of the preceding vertex in the
+    path for that vertex. For example, the vertex identifier at the ith element
+    of the vertex array has a distance from the source vertex of the ith element
+    in the distance array, and the preceding vertex in the path is the ith
+    element in the predecessor array.
+
+    Examples
+    --------
+    >>> import pylibcugraph, cupy, numpy
+    >>> srcs = cupy.asarray([0, 1, 2], dtype=numpy.int32)
+    >>> dsts = cupy.asarray([1, 2, 3], dtype=numpy.int32)
+    >>> weights = cupy.asarray([1.0, 1.0, 1.0], dtype=numpy.float32)
+    >>> resource_handle = pylibcugraph.experimental.ResourceHandle()
+    >>> graph_props = pylibcugraph.experimental.GraphProperties(
+    ...     is_symmetric=False, is_multigraph=False)
+    >>> G = pylibcugraph.experimental.SGGraph(
+    ...     resource_handle, graph_props, srcs, dsts, weights,
+    ...     store_transposed=False, renumber=False, expensive_check=False)
+    >>> (vertices, distances, predecessors) = pylibcugraph.experimental.sssp(
+    ...     resource_handle, G, source=1, cutoff=999,
+    ...     compute_predecessors=True, do_expensive_check=False)
+    >>> vertices
+    array([0, 1, 2, 3], dtype=int32)
+    >>> distances
+    array([3.4028235e+38, 0.0000000e+00, 1.0000000e+00, 2.0000000e+00],
+          dtype=float32)
+    >>> predecessors
+    array([-1, -1,  1,  2], dtype=int32)
+    """
+
+    # FIXME: import these modules here for now until a better pattern can be
+    # used for optional imports (perhaps 'import_optional()' from cugraph), or
+    # these are made hard dependencies.
+    try:
+        import cupy
+    except ModuleNotFoundError:
+        raise RuntimeError("sssp requires the cupy package, which could not "
+                           "be imported")
+    try:
+        import numpy
+    except ModuleNotFoundError:
+        raise RuntimeError("sssp requires the numpy package, which could not "
+                           "be imported")
+
+    if compute_predecessors is False:
+        raise ValueError("compute_predecessors must be True for the current "
+                         "release.")
+
+    cdef cugraph_resource_handle_t* c_resource_handle_ptr = \
+        resource_handle.c_resource_handle_ptr
+    cdef cugraph_graph_t* c_graph_ptr = graph.c_graph_ptr
+
+    cdef cugraph_paths_result_t* result_ptr
+    cdef cugraph_error_code_t error_code
+    cdef cugraph_error_t* error_ptr
+
+    error_code = cugraph_sssp(c_resource_handle_ptr,
+                              c_graph_ptr,
+                              source,
+                              cutoff,
+                              compute_predecessors,
+                              do_expensive_check,
+                              &result_ptr,
+                              &error_ptr)
+    assert_success(error_code, error_ptr, "cugraph_sssp")
```

Comment:
Something to address:  If `cugraph_sssp` (or any of the C API functions) returns an error then `error_ptr` will have a value.  That will need to be freed using `cugraph_error_free`.

Reply:
added FIXME to the util function that handles those in next commit.